gplugins/design_recipe/DesignRecipe.py

```diff
@@ -84,9 +154,332 @@ def eval_dependencies(self, force_rerun_all=False) -> bool:
+
+        Parameters:
+            dirpath: Directory with pickle file
+
+        Returns:
+            True if results exist, False otherwise.
+        """
+        if isinstance(dirpath, str):
+            dirpath = Path(dirpath)
+        if dirpath is None:
+            results_file = self.dirpath.resolve() / f"{self.prefix}_results.pkl"
+        else:
+            results_file = dirpath.resolve() / f"{self.prefix}_results.pkl"
+        return results_file.is_file()
+
+
+class Simulation:
+    """
+    Represents the simulation object used to simulate GDSFactory devices.
+
+    This simulation object's purpose is to reduce time simulating by recalling
+    hashed results.
+    """
+
+    # the hash of the system last time convergence was executed
+    last_hash: int = -1
+
+    # A dynamic object used to store convergence results
+    convergence_results: Results
+
+    def __init__(
+        self,
+        component: Component,
+        layerstack: LayerStack | None = None,
+        simulation_settings: pydantic.BaseModel | None = None,
+        convergence_settings: pydantic.BaseModel | None = None,
+        dirpath: Path | None = None,
+    ):
+        self.dirpath = dirpath or Path(".")
+        self.component = component
+        self.layerstack = layerstack or get_layer_stack()
+        self.simulation_settings = simulation_settings
+        self.convergence_settings = convergence_settings
+
+        self.last_hash = hash(self)
+
+        # Create directory for simulation files
+        self.simulation_dirpath = (
+            self.dirpath / f"{self.__class__.__name__}_{self.last_hash}"
+        )
+        self.simulation_dirpath.mkdir(parents=True, exist_ok=True)
+
+        # Create attribute for convergence results
+        self.convergence_results = Results(
+            prefix="convergence", dirpath=self.simulation_dirpath
+        )
+
+    def __hash__(self) -> int:
+        """
+        Returns a hash of all state this Simulation contains
+        Subclasses should include functionality-specific state (e.g. convergence info) here.
+        This is used to determine simulation convergence (i.e. if it needs to be rerun)
+
+        Hashed items:
+        - component
+        - layer stack
+        - simulation settings
+        - convergence settings
+        """
+        h = hashlib.sha1()
+        if self.component is not None:
+            h.update(self.component.hash_geometry(precision=1e-4).encode("utf-8"))
+        if self.layerstack is not None:
+            h.update(self.layerstack.model_dump_json().encode("utf-8"))
+        if self.simulation_settings is not None:
+            h.update(self.simulation_settings.model_dump_json().encode("utf-8"))
+        if self.convergence_settings is not None:
+            h.update(self.convergence_settings.model_dump_json().encode("utf-8"))
+        return int.from_bytes(h.digest(), "big")
+
+    def convergence_is_fresh(self) -> bool:
```

Comment:
unsure what this means. is there a reason this isn't `results_are_fresh`? maybe I'm unclear on what exactly 'convergence' means here

Reply:
convergence means that the simulation settings are accurate / stable. Convergence sweeps are run to ensure the simulation settings are accurate / stable. A better description for this method would be `convergence_results_are_fresh`. Convergence results differ from simulation results. The output from convergence results is accurate simulation settings. The output from simulation results is any device characteristic.